### PR TITLE
[JENKINS-42518] Provide lightweight checkout capability for bitbucket…

### PR DESCRIPTION
**Pull request by LuisPiedra**
… to avoid repository clone for multi-branch pipeline jobs

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 1]
```diff
@@ -664,4 +676,61 @@ private String postRequest(String path, NameValuePair[] params) throws IOExcepti
...
         return branches;
     }
+
+    private <T> T parse(String response, Class<T> clazz) throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
+        return mapper.readValue(response, clazz);
+    }
+
+    public Iterable<SCMFile> getDirectoryContent(BitbucketSCMFile parent) throws IOException, InterruptedException {
+        StringBuilder url = new StringBuilder(V1_API_BASE_URL);
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Use V2 as recently V1 as removed

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java [hunk 2]
```diff
@@ -671,4 +702,14 @@ private String deleteRequest(String path) throws IOException {
         return doRequest(request);
     }
 
+    @Override
+    public Iterable<SCMFile> getDirectoryContent(BitbucketSCMFile parent) throws IOException, InterruptedException {
+        throw new IOException("List Directory Content not supported for BitBucket Server");
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: Why no love for Bitbucket server?

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java [hunk 3]
```diff
@@ -23,6 +23,38 @@
  */
 package com.cloudbees.jenkins.plugins.bitbucket.client;
 
+import java.io.ByteArrayInputStream;
```
Comment (casz) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Avoid messing with import order.

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 1 above)
Comment (casz) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Seems like it's the same path:

**Comment by casz** [comment]
Bitbucket server support should be easy :|

https://developer.atlassian.com/static/rest/bitbucket-server/4.5.1/bitbucket-rest.html#idp2872912

REST API way:
```
https://bitbucket.widex.com/rest/api/1.0/projects/BEA/repos/ios/browse/Jenkinsfile?at=da148ccadfbcf66d015710ea4d24c71d1917bbb7
```

```
// 20171101120022
// https://bitbucket.widex.com/rest/api/1.0/projects/BEA/repos/ios/browse/Jenkinsfile?at=da148ccadfbcf66d015710ea4d24c71d1917bbb7

{
  "lines": [
    {
      "text": "/*"
    },
    {
      "text": "  The buildiOSapp step is part of"
    },
    {
      "text": "  the Shared Library is located here:"
    },
    {
      "text": "  https://bitbucket.widex.com/projects/SWD/repos/app-pipeline/browse"
    },
    {
      "text": "*/"
    },
    {
      "text": "@Library('app@ios/stable') _"
    },
    {
      "text": "buildiOSapp()"
    }
  ],
  "start": 0,
  "size": 7,
  "isLastPage": true
}
```

Or RAW view way:
```
https://bitbucket.widex.com/projects/BEA/repos/ios/raw/Jenkinsfile?at=da148ccadfbcf66d015710ea4d24c71d1917bbb7
```

```
/*
  The buildiOSapp step is part of
  the Shared Library is located here:
  https://bitbucket.widex.io/projects/SWD/repos/app-pipeline/browse
*/
@Library('app@ios/stable') _
buildiOSapp()
```

**Comment by LuisPiedra** [comment]
It's just a preliminary version to address my needs... Anyway, it seems @vivek has already started to work on this issue, so I probably won't progress any further.

Probably I shouldn't have opened the PR if I did not really intended to merge it? Sorry, I'm not used to work with PR on opensource projects :(

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java (hunk 1 above)
Comment (LuisPiedra) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/client/BitbucketCloudApiClient.java]: Yep, I started to work on this before V2 support was available in the master branch :(

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java (hunk 2 above)
Comment (LuisPiedra) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/server/client/BitbucketServerAPIClient.java]: It's just I don't have a working instance of BitBucket server to test.

**Comment by casz** [comment]
@vivek maybe you can work of this, this the change seems solid 👍

**Comment by vivek** [comment]
@LuisPiedra thanks. I barely started so this is good. I am going to review your changes. Is it ok if I make changes on your PR branch otherwise I can merge it in to my branch?

**Comment by vivek** [comment]
I guess I will have to pull changes in to my branch if needed. Can’t modify fork as I am not a committer.

**Comment by LuisPiedra** [comment]
@vivek Sure, you can take whatever you want and also make changes in the PR branch.

It has been working perfectly fine for me with BitBucket Cloud, with both pull requests and branches in mercurial repositories. I did only minor tests with git, but it was also working fine.

**Comment by vivek** [comment]
Can someone with commit access close this PR? Its replaced by https://github.com/jenkinsci/bitbucket-branch-source-plugin/pull/78.

**Comment by stephenc** [comment]
Closed by #78

**Pull request closed without merging**